broker/artemis/plugin/src/main/java/org/eclipse/kapua/broker/artemis/plugin/security/context/SecurityContextHandler.java

```diff
@@ -52,87 +85,177 @@ public static SecurityContextHandler getInstance() {
+        printReport(ReportType.DetailedServer, server, caller, connectionId);
     }
 
-    public void setSessionContext(SessionContext sessionContext, AuthResponse authResponse) throws KapuaIllegalArgumentException {
-        setSessionContext(sessionContext);
-        aclMap.put(sessionContext.getConnectionId(), new Acl(sessionContext.getPrincipal(), authResponse));
-        sessionContextMapByClient.put(Utils.getFullClientId(sessionContext), sessionContext);
+    private void printReport(ReportType reportType, ActiveMQServer server, String caller, String connectionId) {
+        StringBuilder builder = new StringBuilder();
+        builder.append("\n").append(REPORT_HEADER).append("\n");
+        switch (reportType) {
+        case Full:
+            appendServerContextReport(builder, server);
+            builder.append(REPORT_SEPARATOR).append("\n");
+            appendSessionInfoReport(builder, server);
+            builder.append(REPORT_SEPARATOR).append("\n");
+            appendDetailedServerContextReport(builder, caller, connectionId);
+            break;
+        case Compact:
+            appendServerContextReport(builder, server);
+            break;
+        case DetailedServer:
+            appendDetailedServerContextReport(builder, caller, connectionId);
+            break;
+        default:
+            break;
+        }
+        builder.append(REPORT_HEADER);
+        logger.info("{}", builder);
+    }
+
+    public void init(ActiveMQServer server) {
+       executorWrapper = new ExecutorWrapper("ServerReport", () -> printReport(server, "ServerReportTask", "N/A"), 60, 30, TimeUnit.SECONDS);
+       executorWrapper.start();
+    }
+
+    public void shutdown(ActiveMQServer server) {
+        if (executorWrapper!=null) {
+            executorWrapper.stop();
+        }
     }
 
-    private SessionContext cleanSessionContext(SessionContext sessionContext) {
-        principalMap.remove(sessionContext.getConnectionId());
-        sessionContextMapByConnection.remove(sessionContext.getConnectionId());
-        if (!sessionContext.isInternal()) {
-            aclMap.remove(sessionContext.getConnectionId());
-            //check for stealing link
-            //remove object only if connection ids match
+    public ConnectionToken updateConnectionTokenOnConnection(String connectionId) {
+        ConnectionToken connectionToken = connectionTokenCache.getAndRemove(connectionId);
+        if (connectionToken==null) {
+            connectionTokenCache.put(connectionId,
+                new ConnectionToken(SecurityAction.brokerConnect, KapuaDateUtils.getKapuaSysDate()));
+        }
+        else {
+            //the disconnect callback is called after the connect so nothing to add to the context
+            //TODO add metric?
+            logger.warn("Connect callback called after the disconnection callback ({} - {} - {})", connectionId, connectionToken.getAction(), connectionToken.getActionDate());
+        }
```

Comment:
I think there's no needings to synchronize this block since is called by the connect and is managing the cache objects by connection id.
There is only one case, from my understanding, when this method could lead to a not good state and is when our disconnect handler is called before this block.
I cannot exclude this option is not possible to have realized but very hard I guess.

Reply:
anyway this method could be private (my fault) since is called by the setSessionContext method in a synchronized block